apps/server/src/modules/oauth-provider/service/oauth-provider.login-flow.service.ts

```diff
@@ -0,0 +1,48 @@
+import { ProviderLoginResponse } from '@shared/infra/oauth-provider/dto';
+import { LtiToolDO } from '@shared/domain/domainobject/ltitool.do';
+import { LtiToolRepo, PseudonymsRepo } from '@shared/repo';
+import { InternalServerErrorException } from '@nestjs/common';
+import { Permission, PseudonymDO, User } from '@shared/domain';
+import { Injectable } from '@nestjs/common/decorators/core/injectable.decorator';
+import { AuthorizationService } from '@src/modules/authorization';
+
+@Injectable()
+export class OauthProviderLoginFlowService {
+	constructor(
+		private readonly ltiToolRepo: LtiToolRepo,
+		private readonly pseudonymsRepo: PseudonymsRepo,
+		private readonly authorizationService: AuthorizationService
+	) {}
+
+	async getPseudonym(currentUserId: string, loginResponse: ProviderLoginResponse): Promise<PseudonymDO> {
+		if (!loginResponse.client.client_id) {
+			throw new InternalServerErrorException('Could not find oAuthClientId in login response to get the pseudonym');
+		}
+
+		const ltiToolDO: LtiToolDO = await this.ltiToolRepo.findByClientIdAndIsLocal(loginResponse.client.client_id, true);
+		const pseudonym: PseudonymDO = await this.pseudonymsRepo.findByUserIdAndToolId(
+			ltiToolDO.id as string,
+			currentUserId
+		);
+		return pseudonym;
+	}
+
+	// TODO N21-91. Magic Strings are not desireable
+	private async isNextcloudTool(loginResponse: ProviderLoginResponse) {
+		if (!loginResponse.client?.client_id) {
+			throw new InternalServerErrorException('Could not find oAuthClientId in login response to get the LtiTool');
+		}
+
+		const ltiToolDO: LtiToolDO = await this.ltiToolRepo.findByClientIdAndIsLocal(loginResponse.client.client_id, true);
+		const isNextcloud: boolean = ltiToolDO.name === 'SchulcloudNextcloud';
+		return isNextcloud;
+	}
+
+	async validateNextcloudPermission(currentUserId: string, loginResponse: ProviderLoginResponse) {
+		const isNextcloudTool: boolean = await this.isNextcloudTool(loginResponse);
+		if (isNextcloudTool) {
```

Comment:
if I remember correctly, in the last version of this function there was an error thrown when it was not a nextcloud tool, I think now it would go through successfully

Reply:
There was no error if it was not a nextcloud tool. There was one when it was a nextcloud tool and the user does not have the permission. The error is now thrown by the checkAllPermissions method.